web/modules/custom/joinup_federation/src/Plugin/pipeline/Step/BrokenReferences.php

```diff
@@ -0,0 +1,177 @@
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $sparql);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->entityFieldManager = $entity_field_manager;
+    $this->rdfSchemaFieldValidator = $rdf_schema_field_validator;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('sparql_endpoint'),
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager'),
+      $container->get('rdf_schema_field_validation.schema_field_validator')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    $blacklist = array_flip($this->getPersistentDataValue('blacklist'));
+    $ids = array_keys($this->getPersistentDataValue('entities'));
+    /** @var \Drupal\rdf_entity\RdfInterface $entity */
+    foreach ($this->getRdfStorage()->loadMultiple($ids, ['staging']) as $id => $entity) {
+      $changed = FALSE;
+      $reference_fields = $this->getAdmsSchemaEntityReferenceFields($entity->bundle(), ['rdf_entity', 'taxonomy_term']);
+
+      foreach ($reference_fields as $field_name => $target_entity_type_id) {
+        /** @var \Drupal\Core\Field\EntityReferenceFieldItemListInterface $field */
+        $field = $entity->get($field_name);
+
+        // Remove references to entities that were blacklisted by the user.
+        if ($blacklist && ($target_entity_type_id === 'rdf_entity')) {
+          $changed |= $this->removeBlacklistedReferences($field, $blacklist);
+        }
+        // Remove references to non-existing taxonomy terms.
+        elseif ($target_entity_type_id === 'taxonomy_term') {
+          $changed |= $this->removeTermsBrokenReferences($field);
+        }
+      }
+
+      if ($changed) {
+        $entity->skip_notification = TRUE;
+        $entity->save();
+      }
+    }
+  }
+
+  /**
+   * Removes the items referencing blacklisted entities from a field.
+   *
+   * @param \Drupal\Core\Field\EntityReferenceFieldItemListInterface $field
+   *   The entity reference field item list.
+   * @param array $blacklist
+   *   The list of blacklisted entity IDs.
+   *
+   * @return bool
+   *   If at least one field item has been removed.
+   */
+  protected function removeBlacklistedReferences(EntityReferenceFieldItemListInterface $field, array $blacklist): bool {
+    $changed = FALSE;
+
+    if (!$field->isEmpty()) {
+      $deltas_to_remove = [];
+      foreach ($field as $delta => $field_item) {
+        if (isset($blacklist[$field_item->target_id])) {
+          $deltas_to_remove[] = $delta;
+        }
+      }
+      if ($deltas_to_remove) {
+        foreach ($deltas_to_remove as $delta_to_remove) {
+          $field->removeItem($delta_to_remove);
```

Comment:
This is not going to work properly if more than one deltas are found for removal. That is because the deltas array is build above and then they are removed here. But if, lets say and array of 2 items exists (index 0 and index 1), and both of them are to be removed, then the removeItem method will first remove index 0, but then only index 0 will remain as the other items are shifted.
I noticed it breaks the functionality when more items are to be removed.

Reply:
This is a good catch. I tested only with one item